sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs

```diff
@@ -65,150 +67,186 @@ public sealed class AvoidExcessiveClassCoupling : ParameterLoadingDiagnosticAnal
+                    var typeDeclaration = (TypeDeclarationSyntax)c.Node;
 
-                    if (classDeclaration.Identifier.IsMissing)
+                    if (typeDeclaration.Identifier.IsMissing)
                     {
                         return;
                     }
 
-                    var allFieldsCoupling = classDeclaration.Members
-                        .OfType<FieldDeclarationSyntax>()
-                        .SelectMany(field => CollectCoupledClasses(field, c.SemanticModel));
+                    var collector = new TypeDependencyCollector(c.SemanticModel, typeDeclaration);
+                    collector.Visit(typeDeclaration);
 
-                    var allPropertiesCoupling = classDeclaration.Members
-                        .OfType<BasePropertyDeclarationSyntax>()
-                        .SelectMany(property => CollectCoupledClasses(property, c.SemanticModel));
+                    var dependentTypes = collector.DependentTypes
+                        .OfType<INamedTypeSymbol>()
+                        .SelectMany(ExpandGenericTypes)
+                        .Distinct()
+                        .Where(IsCountableType)
+                        .ToList();
 
-                    var allMethodsCoupling = classDeclaration.Members
-                        .OfType<BaseMethodDeclarationSyntax>()
-                        .SelectMany(baseMethod => CollectCoupledClasses(baseMethod, c.SemanticModel));
-
-                    var totalClassCoupling = Enumerable.Empty<ITypeSymbol>()
-                        .Union(allFieldsCoupling)
-                        .Union(allPropertiesCoupling)
-                        .Union(allMethodsCoupling);
-                    var classCouplingCount = FilterAndCountCoupling(totalClassCoupling);
-
-                    if (classCouplingCount > Threshold)
+                    if (dependentTypes.Count > Threshold)
                     {
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, classDeclaration.Identifier.GetLocation(),
-                            classCouplingCount, Threshold));
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, typeDeclaration.Identifier.GetLocation(),
+                            typeDeclaration.Keyword.ValueText, dependentTypes.Count, Threshold));
                     }
                 },
-                SyntaxKind.ClassDeclaration);
+                SyntaxKind.ClassDeclaration,
+                SyntaxKind.StructDeclaration,
+                SyntaxKind.InterfaceDeclaration);
         }
 
-        private static IEnumerable<ITypeSymbol> CollectCoupledClasses(FieldDeclarationSyntax field,
-            SemanticModel model)
+        private static bool IsCountableType(INamedTypeSymbol namedType) =>
+            namedType.TypeKind != TypeKind.Enum &&
+            !namedType.IsAny(ignoredTypes);
+
+        private static IEnumerable<INamedTypeSymbol> ExpandGenericTypes(INamedTypeSymbol namedType)
```

Comment:
Some documentation of the intention would be useful e.g. what is expected for _Dictionary<string, MyClass>_?